# Homework 1

**Importing labs**

In [1]:
import pandas as pd

In [49]:
import numpy as np

In [17]:
pd.__version__

'3.0.6'

1. Pandas version is 3.0.6

**Importing data**

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv")

In [3]:
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


**How many records are in the dataset?**

In [5]:
df.shape

(10000, 11)

2. Our dataset has 10 000 records 

**How many fuel types are presented in the dataset?**

In [6]:
df.fuel_type.nunique()

3

3. In the dataset three different fuel types

**How many columns in the dataset have missing values?**

In [12]:
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

4. Only two columns have missing values 

In [15]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(41.2)

The maximum value for Asian region is 41.2


6. **Find the median value of the horsepower column in the dataset.**<br>
**Next, calculate the most frequent value of the same horsepower column.**<br>
**Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.**<br>
**Now, calculate the median value of horsepower once again.**<br>

    **Has it changed?**<br>
    **Yes, it increased**<br>
    **Yes, it decreased**<br>
    **No**<br>

In [16]:
df.horsepower.median()

np.float64(254.0)

a. The median is 254

In [24]:
df.horsepower.value_counts().idxmax()

np.float64(252.0)

b. The most frequent value is 252

In [26]:
df.horsepower = df.horsepower.fillna(252)

In [28]:
df.horsepower.median()

np.float64(252.0)

c. The median value has decreased, now it's 252

7. **Select all the cars from Asia**<br>
**Select only columns vehicle_weight and model_year**<br>
**Select the first 7 values**<br>
**Get the underlying NumPy array. Let's call it X.**<br>
**Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.**<br>
**Invert XTX.**<br>
**Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].**<br>
**Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.**<br>
**What's the sum of all the elements of the result?**<br>

In [29]:
asia = df[df['origin'] == 'Asia']

In [32]:
asia = asia[['vehicle_weight', 'model_year']]

In [37]:
asia = asia.iloc[0:7]

In [38]:
asia

,vehicle_weight,model_year
2,4240,1996
6,4310,1992
8,4230,1997
10,4420,1978
15,3950,2007
16,4750,2021
22,4450,2019


In [41]:
X = asia.values

In [42]:
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

In [44]:
XTX = X.T @ X

In [45]:
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

In [54]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [55]:
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [56]:
r = np.linalg.inv(XTX) @ X.T

In [57]:
r

array([[-2.50082897e-04, -1.65786806e-05, -2.86196661e-04,
         3.99964853e-04, -1.18162927e-03,  1.10298508e-03,
         2.25354894e-04],
       [ 6.12974614e-04,  1.06954785e-04,  6.91249245e-04,
        -7.95968736e-04,  2.63152339e-03, -2.31749941e-03,
        -4.16221392e-04]])

In [58]:
w = r @ y 

In [59]:
w

array([0.13644777, 0.2327492 ])

In [60]:
w.sum()

np.float64(0.36919696904925486)